backend/src/index.ts

```diff
@@ -89,44 +137,55 @@ const processSourceBlocks = (target: Target) => async (source: Source) => {
     const master = new PoolSigner(targetApi.registry, config.accountSeed, 1);
 
     if (args.length && (args[0] === 'archive')) {
-      const archives = await Promise.all(config.archives.map(async ({ path, url }) => {
+      config.archives.map(async ({ path, url }) => {
         const chain = await getChainName(url);
         const signer = new PoolSigner(
             target.api.registry,
             `${config.accountSeed}/${chain}`,
             SIGNER_POOL_SIZE,
         );
+        // TODO: Do not send balance
         await target.sendBalanceTx(master, signer.address, 1.5);
         const feedId = await target.getFeedId(signer);
 
-        return new ChainArchive({
+        // TODO: Remove signer and feedId from arguments, maybe something else too
+        const archive = new ChainArchive({
           path,
-          chain,
-          feedId,
           logger,
-          signer,
-          state,
         });
-      }))
 
-      archives.forEach(async archive => {
         let lastBlockProcessingReportAt = Date.now();
-        let processedBlocks = 0;
-
-        let nonce = (await target.api.rpc.system.accountNextIndex(archive.signer.address)).toBn();
-        for await (const blockData of archive.getBlocks()) {
-          target.sendBlockTx(blockData, nonce);
-          nonce = nonce.add(new BN(1));
-
-          processedBlocks++;
 
-          if (processedBlocks % REPORT_PROGRESS_INTERVAL === 0) {
-            const now = Date.now();
-            const rate = (Number(REPORT_PROGRESS_INTERVAL) / ((now - lastBlockProcessingReportAt) / 1000)).toFixed(2);
-            lastBlockProcessingReportAt = now;
+        let nonce = (await target.api.rpc.system.accountNextIndex(signer.address)).toBigInt();
+        const lastProcessedString = await state.getLastProcessedBlockByName(chain);
+        const lastProcessedBlock = lastProcessedString ? parseInt(lastProcessedString, 10) : 0;
 
-            logger.info(`Processed downloaded ${blockData.chain} block ${blockData.metadata.number} at ${rate} blocks/s`);
+        let lastTxPromise: Promise<void> | undefined;
+        for await (const [blocksBatch, lastBlockNumber] of readBlocksInBatches(archive, lastProcessedBlock)) {
+          if (lastTxPromise) {
+            await lastTxPromise;
           }
+          lastTxPromise = (async () => {
+            try {
+              const txHash = await target.sendBlocksBatchTx(feedId, signer, blocksBatch, nonce);
+              nonce++;
+
+              logger.debug(`Transaction included: ${polkadotAppsUrl}${txHash}`);
```

Comment:
this URL won't work for transactions, you can only query blocks

Reply:
It is a bad variable name, turns out it is a block hash